### PR TITLE
Parse NIRISS data files by header instead of name

### PR DESCRIPTION
**Pull request by ojustino**
### Description

I updated `mos_niriss_parser()` to sort files into categories (image, 1d spectrum, etc.) by examining their FITS headers instead of matching their filenames to a pattern. I chose distinctive keywords based on the headers of the current NIRISS example's data files. The `'DATAMODL'` key is particularly useful.

I can run the NIRISS example in full and don't see any glaring errors.

Open questions include:
- Are these the best ways to sort NIRISS' data files into the categories Mosviz desires?
- ~~What is the expected structure of the .ecsv source catalog? Two columns for RA and Dec or one with them in a tuple?~~
- ~~Should we still allow users to provide an `obs_label` argument to select a subset of NIRISS files in a directory? If so, what keyword is best? I left some examples in the docstring.~~

[🐱](https://jira.stsci.edu/browse/JDAT-1619)

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:457]: I am a bit lost here. `file_lists['Source Catalog']` is still a list of filepaths, right? Why would column name be in it? Don't you have to parse the actual file into `astropy.table.Table` first? Did I miss something?

`pop` and `extend` is a little confusing, I would prefer a simpler if-else using `+` on list, but that is just a matter of personal taste.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:478]: So, are the source catalogues allowed to be mixed among "new" and "old" styles?

* If yes: `cat_fields` should be recomputed at every iteration here.
* If no: There is no need to recompute `len(cat_fields)` at every iteration.

**Review by pllim — COMMENTED** [review]
The FITS header logic is too specific. I wonder if you need to check for other stuff like INSTRUME, TELESCOP, and CAL_VER. I'll leave that to your good judgment.

Though on the other hand, I am starting to wonder, does `jwst.datamodel` not provide a way to understand their outputs in a more abstract way? This feels a little like reverse engineering.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:402]: themoreyouknow.gif

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:393]: This is one of the discussion questions. I'm not sure which of these labels (or others) should be used for this argument.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:405]: I wasn't planning to support recursion here. This is why I also start by checking whether the path is a directory.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:421]: I would prefer to get a `None` here if "FILTER" is missing so that "X" isn't mistaken as legitimate output by someone in the future.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:442]: This case is here for my own testing purposes. I'll remove it entirely once the code is finalized since it would already be the last step of any iteration that reached it.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:457]: You're right; this is my mistake. (The pop and extend is a placeholder.) The problem here comes from the comment above this code about the change in the catalog table's columns. I guess the only place to test which column names are valid is inside `_fields_from_ecsv()`.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:478]: I tried to support both scenarios because I don't know if the old one is no longer valid. Whom should I ask about that?

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [reply jdaviz/configs/mosviz/plugins/parsers.py:393]: Perhaps a question for pipeline team. There must be some logic to it but I cannot find that easily at https://jwst-pipeline.readthedocs.io/ . Maybe you can. 🤷

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [reply jdaviz/configs/mosviz/plugins/parsers.py:478]: 🤷 I'll have to defer to @javerbukh or @rosteen .

**Comment by ojustino** [comment]
@plim, checking INSTRUME and TELESCOP makes sense to me, but I assumed that would take place before we reach `mos_niriss_parser()`. Perhaps CAL_VER can help with the source catalog column name issue, though. I can do some more digging to see if `jwst.datamodels` is getting anything else that would be helpful here.

**Comment by pllim** [comment]
> INSTRUME and TELESCOP

Oh, right. This is NIRISS parser, not the generic one. Forget I said anything. 😆

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:457]: Turns out `"skycoord.ra"` and `"skycoord.dec"` are valid. I thought they were actual column names instead of variables set at the head of the .ecsv file. I'll change code related to this `cat_fields` back to its previous state.

**Comment by javerbukh** [comment]
This looks good, although the data @havok2063 provided contains a `json` file rather than an `ecsv`, so I can't test this on that data. Based on the name of this PR though, handling `json` files may be out of scope. I'll do another review once the code style checks and merge conflicts are fixed.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:478]: This question doesn't apply anymore given the resolution of the conversation just before this one. I would instead go to the questions in the pull request's initial post.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:445]: I actually think this is more of a question for the team -- how, if at all, should `jdaviz` users be able to limit which files in a directory are loaded into Mosviz? Should the burden be on the user to only include relevant files in their data directory, or should `mosviz.load_data()` provide a way to only load files with some desired characteristic?

**Comment by havok2063** [comment]
@javerbukh There are some ecsv files in the 787 data directory I uploaded, at least one in the L3/t directory.  I'm not sure if this what you're looking for.   If you can give me some information I can try and track down the relevant files and add them to Box.  And there's an association pool csv file in the `asn` directory.  But yeah `json` files I think are the main format tracking associations.   There's more info https://jwst-pipeline.readthedocs.io/en/latest/jwst/associations/level3_asn_technical.html#example-association and https://jwst-pipeline.readthedocs.io/en/latest/jwst/associations/overview.html

**Comment by ojustino** [comment]
This is a separate question from the ones at the beginning of this pull request, but have we handled NIRISS association files in the past?

**Comment by pllim** [comment]
> have we previously handled NIRISS association files

AFAIK, no.

**Comment by rosteen** [comment]
Have you rebased to main since the hotfix PR was merged? I think that may fix the 3.8 test failure.

**Comment by ojustino** [comment]
@rosteen, I started that at the end of the day yesterday but didn't push. Let's see how it goes...

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (larrybradley) [reply jdaviz/configs/mosviz/plugins/parsers.py:444]: @ojustino I spoke with @orifox and we agree that this is out of scope for now.  We may add some sort of file filter in the future, but not right now.

**Comment by larrybradley** [comment]
@ojustino Regarding your question:
"What is the expected structure of the .ecsv source catalog? Two columns for RA and Dec or one with them in a tuple?"

If you read an `.ecsv` source catalog file with `astropy.Tables`, you will get a single column containing an astropy `SkyCoord` object (https://docs.astropy.org/en/stable/api/astropy.coordinates.SkyCoord.html), which contains both the RA/Dec and the coordinate frame.  For JWST data, I think the coordinate frame is always ICRS.  If you want to get just the RA or Dec from a `SkyCoord`, you can use the `ra` and `dec` attributes.

Example usage:

```python
>>> from astropy.coordinates import SkyCoord
>>> coord = SkyCoord.from_name('m51')
>>> coord
<SkyCoord (ICRS): (ra, dec) in deg
    (202.469575, 47.19525833)>
>>> coord.ra
<Longitude 202.469575 deg>
>>> coord.dec
<Latitude 47.19525833 deg>
>>> coord.icrs.ra
<Longitude 202.469575 deg>
>>> coord.icrs.dec
<Latitude 47.19525833 deg>
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:444]: @larrybradley OK, I'll remove this argument and perhaps create an issue for it.

**Comment by ojustino** [comment]
Thanks, @larrybradley, I figured out the `.ecsv` column issue in one of the resolved comment chains. The NIRISS parser doesn't currently use `QTable`, but the source catalog file's header assigns `ra` and `dec` attributes to its `sky_centroid` column to serve a similar purpose as in your example.

I'll need to find whether the source catalog are the same in the newer pipeline files we got from @havok2063. I'm trying to mock the structure of a level 3 NIRISS data directory from the files there and will see how the parser fares if I'm able to do that.

**Comment by ojustino** [comment]
I'll leave a progress report since I'm off today:

### 1
A discussion with @havok2063 revealed that our example NIRISS data directory from `MosvizNIRISSExample.ipynb` may be a mix of level 2b and 3 files instead of level 3 alone. Specifically, the 2D spectrum files' `DATAMODL` value of "MultiSlitModel" is not found in any of the spectral level 3 data files we have in the `1.3.1` folder for observation jw00787 in Box.

Instead, it seems like the C and R dispersion directions have their data merged into one file with a `DATAMODL` value of "MultiCombinedSpecModel." I will check other level 3 data from the MAST portal to see if that's consistently the case, but **@camipacifici, does this sound correct for NIRISS?**

If so, I need to figure out how to split the science data by dispersion direction. It's not immediately obvious to me looking at the headers.

### 2
The `.ecsv` source catalog uses 'label' instead of 'id' to distinguish sources in our data from pipeline version 1.3.1. (The NIRISS data directory we've been using was made with version 0.17.1 of `jwst`.) **Should there still be support for the 'id' column name?**

Also, `main`'s construction of the NIRISS parser code depends on learning the pupil used in an observation from the source catalog's filename. This would work with default pipeline files, but we can't guarantee that users won't rename them. James says we can't get the pupil from the catalogs' headers, so I'm working on a different approach using the `SCATFILE` key from the other files' FITS headers.

**Comment by camipacifici** [comment]
@ojustino About the level 3 products for NIRISS: yes, the pipeline combines the R and C grisms by default (once done, you cannot split the data back to R and C, I believe), but this is really not a desired behavior, so people will likely rerun stage 3 of the pipeline with two separate association files, one with grism R and one with grism C. I think we should be ready to parse the files in both cases (files coming from the standard pipeline with R and C combined and files coming from the custom pipeline with R and C separated).

**Comment by havok2063** [comment]
I think depending on the observing mode, or what was specified in the observing proposal, the standard pipeline can produce separate data files for each grism.  There are at least some example observations from rehearsals that's produced a mix of `gr150c`, `gr150r` and `gr150c-gr150r` files.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ibusko) [reply jdaviz/configs/mosviz/plugins/parsers.py]: I tend to agree with @ojustino in that we shouldn't rely on exceptions to do useful stuff. Conditional code would be better.

**Comment by ojustino** [comment]
After a rebase, this branch may now depend on the developer version of `specutils` to avoid issues solved in https://github.com/astropy/specutils/pull/856.

**Comment by rosteen** [comment]
I'm trying to get specutils 1.4 out this morning with that fix, so hopefully it won't be a problem for long.

**Comment by pllim** [comment]
Well... we could temporary pin `setup.cfg` here to use specutils at dev hash but I'll defer to @rosteen since he maintains specutils.

**Comment by pllim** [comment]
(oh ops... I didn't see his reply when I commented...)

**Comment by ojustino** [comment]
Handling `c1d` files looks like a moot point after the decision to only support level 2 files for the time being.

Whenever we schedule a meeting on NIRISS, my blockers are:
- ...not having seen level 2 files for a more recent pipeline build, specifically `x1d` (one for each dispersion direction), `cal` (one for each dispersion direction), `i2d`, and a `.ecsv` source catalog.
- ...not knowing how to drop dependence on filename for the source catalog since its header doesn't contain pupil information.

**Comment by camipacifici** [comment]
If Bryan is not already working on a newer set of simulations, I can commit to create that.

About the ecsv, maybe it is on the user to associate the correct ecsv with the relative i2d, cal, and x1d file. Alternatively, the information is in the actual association file the pipeline uses. The user could provide that file too in the folder for loading.

**Comment by ojustino** [comment]
Our `jwst` 1.3.1 example files from @havok2063 won't work because they have no 2D spectra associated with an `.ecsv` source catalog. The NIRISS parser depends on finding RA/Dec/image info in 2D spectrum headers.

I got @camipacifici's example files from `jwst` 1.3.2 working in a way that preserves compatibility with our previous example files. I still run into the same issue of depending on the source catalog's file name to learn its associated pupil, though.

I would find it helpful if we clarified our expectations for how users' file directories should look. For example, should we say we only support level 2 NIRISS data products? Should we not support source catalogs with custom file names? What happens if they have multiple source catalogs or images (like `i2d` and `cal`) in one directory?

**Comment by havok2063** [comment]
I would strongly suggest that Mosviz be made to work with the standard pipeline output first, before any custom files.  I think most users will download the data from MAST via the Portal or astroquery and attempt to load it into Mosviz.  A subset of users will rerun the standard pipeline themselves.  Another subset of users will customize the pipeline, rerun it, and rename files.  NIRISS doesn't produce resampled 2d spectral files, but I think the relevant information you need can be recovered from the existing files.  If there are certain level 2 output files that are required for Mosviz to fully understand the inputs, and they aren't found or provided by the user, e.g. only a directory of level 3 products given, then an error should be issued to the user.    

Regarding the example dataset I provided, it's only a subset of the pipeline output.  Some of the files you need may be missing.  The full directory is ~6 GB and I don't know about uploading all of that to Box.  I'm happy to provide some instructions on how to search for a full set of products from the Portal.    

As for the directories, I think there are 5 directory structures Mosviz should be designed to work with.
- The MAST isilon directory system - standard filenames in an specifically organized directory
- The Portal bundle download directory - standard filenames with a less-specific but organized directory
- The astroquery product download - (I think) standard filenames in a flattened directory 
- The JWST pipeline directory output -  (I think) standard filenames in a flattened directory
- A custom user supplied directory - could be custom filenames but should be a flattened directory, for the ease of it.

**Comment by camipacifici** [comment]
Getting to the bullets in a second. For the docs:
When you mention the files can you please add that the i2d and cat are level 3 products from the Imaging pipeline (calwebb_image3), while the cal and x1d files are level 2 products from the Spectroscopic pipeline (calwebb_spec2)?

**Comment by camipacifici** [comment]
For completeness (since we only talked offline about this), let's ask the users to provide only one catalog for now, independent of the filter used for source detection.

**Review by duytnguyendtn — COMMENTED** [review]
Overall, this process looks a lot more robust than the filename assumptions; glad to see this get revived! I have a few thoughts to help get this moving forward:

Currently the nirspec parser still makes filename assumptions to identify the types. I wonder if this same logic can be extended in some way to the nirspec parser as well. Regardless, I'd really like to see this built, not as a NIRISS parser, but a generic "file header parser" that gets passed NIRISS specific details. 

Therefore, for readability and extensibility, I think it would be nice to start breaking up the functionality here and extract out the NIRISS-specific stuff as arguments to more generic methods.

For example, I could see the datamodel parsing being separated out into a `identify_files_by_datamodl()` which takes a list of files, and a dict that matches which types have which keyword (`{'1D:' 'MultiSpecModel', '2D': MultiSlitModel', 'Image': 'ImageModel'}`)

Or, you could even abstract it further and make it `identify_files_by_header()` and allow multiple keywords to be passed, like `{'1DR': {'datamodl': 'MultiSpecModel', 'dispersion': 'C'}, '1DC': ...}`

Hopefully if we accomplish this, we can start to think about whether it's possible to unify the two parsers to use the same strategies, and just pass through their instrument-specific keywords to these instrument-agnostic, general strategies

What do you think?

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (javerbukh) [inline jdaviz/configs/mosviz/plugins/parsers.py:626]: I found where [this](https://github.com/spacetelescope/jdaviz/commit/505edfe0128a61d46cee1599c436a76d1b011285#diff-9dac277cb692b00ef5b97d4aedcc083497fda675acdec85547c34a7ec11a3c1bL440) used to be used. This is back when we would add data from the SCI extension of the 2d spectra, but for some reason the WCS information was in another file (image maybe? Can't remember). I think this can safely be removed now since we load data another way.

**Comment by javerbukh** [comment]
@duytnguyendtn I think that sounds like a good idea but I would prefer that to be a separate ticket that follows up this work. Generalizing this work would be non-trivial and I would prefer to merge this PR ASAP if it meets the original criteria.

**Comment by ojustino** [comment]
@duytnguyendtn, I agree with you that Mosviz would be better off with a general parser, but I second @javerbukh's position that this pull request is not the best place to do that.

The ticket this work addresses is specific to NIRISS, and expanding the scope to also include NIRSpec will make this PR's conversational thread even more circuitous than it already is. It would be nice to have separate records for each discussion.

I do endorse the idea of eventually separating `DATAMODL` parsing into its own function, but again, I favor keeping the focus on `mos_niriss_parser()` for this review.

**Comment by duytnguyendtn** [comment]
> expanding the scope to also include NIRSpec will make this... even more circuitous

Totally agree; I didn't mean to suggest we do this unification work here; that should definitely be a separate effort. I just mean we should keep this in mind when we proceed with this PR. In that vein, here's maybe a more direct way of saying what I said above:

The NIRISS parser has always been a monolithic code block due to the complexities of the parsing structure. For both immediate maintainability and readability, it might we worth it to start breaking it up so that the code is a little bit more approachable? Even if right now, we only break the parser up into a bunch of inner/nested functions and we still hard code the NIRISS parts, breaking up the code now with the understanding we might want to abstract it down the line could give us an easier starting point of "which code blocks to rip out", while it's still fresh in our heads

But if this is still too much right now, I'm willing to concede and review as is

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:78]: Not your diff but this looks like leftover debugging?
```suggestion
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:690]: You might want to keep this line to inherit metadata from FITS header? Maybe this is why your CI is failing? Hard to tell from the diff.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:78]: Yeah, I was wondering why it should print _and_ broadcast. I'll accept the deletion since we had the same thought.

**Comment by ojustino** [comment]
@duytnguyendtn, I think that's a fair compromise for this pull request that keeps an eye toward the future. I'll split out the file identification portion of `mos_niriss_parser()` into its own class method as a start.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:690]: This is was indeed the issue. The failing lines pass now for me locally, so let's hope it's reflected in the newest CI results.

Comment (pllim) [reply jdaviz/configs/mosviz/plugins/parsers.py:690]: Yup, tests pass now. 😸

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:610]: Maybe `defaultdict` would simplify this setup, but depends on what you are trying to do, I guess.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:678]: Is this print intentional? Or should we take advantage of debug snackbar logging?

**Review by pllim — COMMENTED** [review]
I don't know the file format enough to do a thorough review. Just a couple of minor comments.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:610]: You actually taught me about `defaultdict` closer to the beginning of this pull request. I stopped using it because I didn't like the implication of `defaultdict(list)` for the source catalog key. The function only accepts one source catalog file, but a list can have more than one entry. I made that key's value a string instead.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (ojustino) [reply jdaviz/configs/mosviz/plugins/parsers.py:678]: The print statements are grandfathered in and weren't my addition. Logging the process as cell output as opposed to a snackbar message has made debugging easier, but that may not be the best reason to continue the practice.

Comment (pllim) [reply jdaviz/configs/mosviz/plugins/parsers.py:678]: OK, no prob. We can defer this.

Comment (javerbukh) [reply jdaviz/configs/mosviz/plugins/parsers.py:678]: It was asked for by @camipacifici when the parser was first created.

**Review by javerbukh — APPROVED** [review]
Works great, thank you for putting in all the effort to resurrect this PR!

**Review by duytnguyendtn — APPROVED** [review]
NIRISS notebook loads well! Definitely appreciate the datamodl chunk being separated out; makes it more readable. Thanks @ojustino!

**Pull request merged**
